### DISCUSSION TITLE
How to define a new platform and target kind?

### DISCUSSION THREAD

**Discussion by davidchisnall** [xmake-io/xmake#2763] [Q&A]
The docs on defining a new toolchain are very clear but even after spending a morning reading the code I'm not clear how to define:

 - A new platform
 - A new target kind

I've tried using `platform("my platform name")` but this errors out that `platform` is nil.  I can use `set_kind("any string")` in the target definition but I can't then see how to define rules for this kind of target (for example, inheriting the existing C/C++ compile rules but providing a different link step).

Edit: I have found the undocumented top-level `add_platformdirs` command and now my platform is being loaded.  It appears that `set_os` is required here, because link steps are silently fail to run if it is not called.

**Comment by waruqi** [comment]
You can't define them directly, but you can use any platform name.

```lua
if is_plat("myplat") then
    add_defines("xxx")
end
```

```console
xmake f -p myplat
xmake
```

you can also use custom rule.

```lua
rule("myplat")
    on_load(function (target)
        if target:is_plat("myplat") then
            target:add("defines", "xxx")
        end
    end)
rule_end()

target("test")
    add_rules("myplat")
    -- ...
```

```console
xmake f -p myplat
xmake
```

and you can put your platform rules to `myplatforms/xmake.lua` and use `includes("myplatforms")` to import them.

**Reply by davidchisnall** [reply]
It appears that if I put them in `platforms/{my platform}/xmake.lua` and then do `add_platformdirs("platforms")` then my platform is loaded and applied correctly.

For the custom target kinds, I can do `set_kind("{whatever"})`, but then none of the default rules run.  Is there a mechanism for me to, for example, import all of the generic compile rules for that kind but provide custom linking rules?

**Reply by waruqi** [reply]
> It appears that if I put them in `platforms/{my platform}/xmake.lua` and then do `add_platformdirs("platforms")` then my platform is loaded and applied correctly.
> 

The platform() interfaces are not yet open for customisation, so I don't have them documented at the moment. As the platform api design is not fully stable yet, I may change them at any time.

Although the add_platformdirs interface is available, I don't recommend that you use it now, please try to customise the platform using the custom rule pattern I mentioned above.

> For the custom target kinds, I can do set_kind("{whatever"}), but then none of the default rules run. Is there a mechanism for me to, for example, import all of the generic compile rules for that kind but provide custom linking rules?

We can't customize kind, you can only use binary/static/shared/headeronly/phony, you should use custom rules to extend kind.

for example: `mykind` rule

```lua
rule("mykind")
    on_load(function (target)
         target:set("kind", "binary")
         target:add("defines", "xxx")
    end)
rule_end()

target("test")
     add_rules("mykind")
```

see https://github.com/xmake-io/xmake/blob/1acf248c04f6ccf89bc3dfed5f37bd3b3a1617f2/xmake/rules/qt/xmake.lua#L40

https://github.com/xmake-io/xmake/blob/1acf248c04f6ccf89bc3dfed5f37bd3b3a1617f2/xmake/rules/mdk/xmake.lua#L34

**Reply by waruqi** [reply]
You can always use custom rule to achieve everything you need, including custom platforms and kind

**Reply by davidchisnall** [reply]
Thanks!  `add_deps` doesn't seem to be documented in the custom rules section (it is used in one example, but I missed it).  It seems to be the thing that was missing for me to be able to do most of the things that I want.  

The other thing that I didn't really understand from the docs is that each of the `set_{thing}` methods in the target scope corresponded to `target:set("{thing}", ...)` in the script scope.  It would be good if the docs could make that explicit.
